### PR TITLE
Add docs for GraphQL Federation

### PR DESCRIPTION
**Pull request by tuxmachine**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[X] Other... Please describe: Documentation
```

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

## Other information

Documents the Federation module from https://github.com/nestjs/graphql/pull/455

### CODE REVIEW

File: content/graphql/federation.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+  id: ID!
+  name: String!
+}
+
+extend type Query {
+  getUser(id: ID!): User
+}
+```
+
+Our resolver has one extra method: `resolveReference`. It'll called by the Apollo Gateway whenever a related resource requires a User instance. We'll see an example of this in the Posts service later on. Please note the `@ResolveReference` decorator.
```
Comment (dolphub) [inline content/graphql/federation.md]: grammatical error

```It'll called```

**Comment by kamilmysliwiec** [comment]
cc @johnbiundo can you review when you have some time? it just got unlocked :)

File: content/graphql/federation.md (hunk 1 above)
Comment (tuxmachine) [reply content/graphql/federation.md]: Thanks, fixed

File: content/graphql/federation.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
+### Federation
+
+[Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) offers a means of splitting your monolith GraphQL server into independent microservices. It consists of two components: A gateway and one or more federated microservices. Each microservice holds part of the schema and the gateway merges the schemas into one single schema that can be consumed by the client.
+
+To quote the [Apollo docs](https://blog.apollographql.com/apollo-federation-f260cf525d21), Federation is designed with these core principles
```
Comment (johnbiundo) [inline content/graphql/federation.md]: these core principles 
->
these core principles:

File: content/graphql/federation.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+- **Building a graph should be declarative.** With federation, you compose a graph declaratively from within your schema instead of writing imperative schema stitching code.
+- **Code should be separated by concern, not by types.** Often no single team controls every aspect of an important type like a User or Product, so the definition of these types should be distributed across teams and codebases, rather than centralized.
+- **The graph should be simple for clients to consume.** Together, federated services can form a complete, product-focused graph that accurately reflects how it’s being consumed on the client.
+- **It’s just GraphQL, using only spec-compliant features of the language.** Any language, not just JavaScript, can implement federation.
+
+> warn **Note** Apollo Federation currently does not support subscriptions, and only the "Schema first" approach is currently supported due to limitations with the decorators not yet supporting GraphQL directives<sup>[1](https://github.com/MichalLytek/type-graphql/issues/351)</sup>
+
+In the next example, we'll set up a demo application with a gateway and two federated endpoints; a user- and posts service.
```
Comment (johnbiundo) [inline content/graphql/federation.md]: endpoints; a user- and posts service.
->
endpoints: a user service and a posts service.

File: content/graphql/federation.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+#### Federated example: Users
+
+First make sure you install the optional dependency for federation: `npm install --save @apollo/federation`.
```
Comment (johnbiundo) [inline content/graphql/federation.md]: remove "make sure"

File: content/graphql/federation.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+The User service has a simple schema. Note the `@key` directive, it tells the Apollo query planner that a particular instance of User can be fetched if you have its `id`. Also note that we `extend` the Query type.
```
Comment (johnbiundo) [inline content/graphql/federation.md]: The User service has a simple schema. Note the `@key` directive, it tells the Apollo query planner that a particular instance of User can be fetched if you have its `id`. Also note that we `extend` the Query type.
->
The User service has a simple schema. Note the `@key` directive: it tells the Apollo query planner that a particular instance of `User` can be fetched if you have its `id`. Also note that we extend the Query type.

File: content/graphql/federation.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  @ResolveReference()
+  resolveReference(reference: { __typename: string; id: string }) {
+    return this.usersService.findById(reference.id);
+  }
+}
+```
+
+Finally, we hook it everything up in a module together with a `GraphQLFederationModule`. This module accepts the same options as the regular `GraphQLModule`.
```
Comment (johnbiundo) [inline content/graphql/federation.md]: we hook it everything
->
we hook everything

File: content/graphql/federation.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+  id: ID! @external
+  posts: [Post]
+}
+
+extend type Query {
+  getPosts: [Post]
+}
+```
+
+Our resolver has one method of interest here: `getUser`. It returns a reference containing `__typename` and any additional properties your application needs to resolve the reference, in this case only an `id`. The `__typename` is used by the GraphQL Gateway to pinpoint the microservice responsible for the User type and request the instance; The User service discussed above will be called on the `resolveReference` method.
```
Comment (johnbiundo) [inline content/graphql/federation.md]: request the instance; The 
->
request the instance. The

File: content/graphql/federation.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  @ResolveProperty("user")
+  getUser(@Parent() post: Post) {
+    return { __typename: "User", id: post.userId };
+  }
+}
+```
+
+The Posts service has virtually the same module, but is included below for the sake of completeness
```
Comment (johnbiundo) [inline content/graphql/federation.md]: completeness
->
completeness:

File: content/graphql/federation.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    })
+  ],
+  providers: [PostsResolvers]
+})
+export class AppModule {}
+```
+
+#### Federated example: Gateway
+
+First make sure you install the optional dependency for the gateway: `npm install --save @apollo/gateway`.
```
Comment (johnbiundo) [inline content/graphql/federation.md]: First make sure you install the optional dependency for the gateway: `npm install --save @apollo/gateway`.
->
First install the optional dependency for the gateway: 
```bash
npm install --save @apollo/gateway
```

File: content/graphql/federation.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+          { name: "posts", url: "http://post-service/graphql" }
+        ]
+      }
+    })
+  ]
+})
+export class AppModule {}
+```
+
+> warn **Warning** Apollo recommends that you don't rely on the service discovery in a production environment but use their [Graph Manager](https://www.apollographql.com/docs/graph-manager/federation/) instead
```
Comment (johnbiundo) [inline content/graphql/federation.md]: instead
->
instead.

File: content/graphql/federation.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+      inject: [GRAPHQL_GATEWAY_BUILD_SERVICE],
+    }),
+  ],
+})
+export class AppModule {}
+```
+
+#### Async configuration
+
+Both the Federation and Gateway modules support asynchronous initialisation using the same `forRootAsync` that's documented in [Quick start](/graphql/quick-start#async-configuration)
```
Comment (johnbiundo) [inline content/graphql/federation.md]: initialisation 
->
initialization

**Comment by tuxmachine** [comment]
Suggestions accepted :) thank you! 

You're right about not including the `.html` file. It's ignored by default so I have no idea how it ended up in there.

**Comment by gperdomor** [comment]
This need more work or can be merged?

**Comment by gperdomor** [comment]
ping @kamilmysliwiec

**Comment by tuxmachine** [comment]
@jarvisuser90 Thank you!

**Comment by kamilmysliwiec** [comment]
Thanks!

**Pull request merged**